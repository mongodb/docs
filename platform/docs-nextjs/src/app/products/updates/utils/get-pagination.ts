/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
