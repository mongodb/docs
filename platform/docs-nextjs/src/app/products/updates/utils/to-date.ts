export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
