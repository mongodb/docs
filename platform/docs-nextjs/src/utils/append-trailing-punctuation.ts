import { ASTNode, ReferenceNode, RefRoleNode } from "@/types/ast";
import { isReferenceNode, isRefRoleNode, isTextNode } from "@/types/ast-utils";

// Append dangling punctuation to preceding Link's text
export const appendTrailingPunctuation = (nodes: ASTNode[]) => {
  const truncatedNodes = [];

  for (let i = 0; i < nodes.length; i++) {
    const currNode = nodes[i];
    const nextNode = nodes[i + 1];
    const hasDanglingSibling =
      isTextNode(nextNode) &&
      nextNode.value.length === 1 &&
      !nextNode.value.match(/\s/);
    // TODO: if issues arise, instead of matching for non-whitespace, we might consider checking FOR certain singular punctuation
    // ie: regex match for  .?!,:;)*"']}`

    if ((isReferenceNode(currNode) || isRefRoleNode(currNode)) && hasDanglingSibling) {
      const copyOfNodeWithDeepCopiedChildren = deepCopyNodesChildren(currNode);
      copyOfNodeWithDeepCopiedChildren.children.push(nextNode);
      truncatedNodes.push(copyOfNodeWithDeepCopiedChildren);
      i++;
      continue;
    }
    truncatedNodes.push(currNode);
  }
  return truncatedNodes;
};

// Make a copy of node with a deep copy of the new child node
function deepCopyNodesChildren(refNode: ReferenceNode | RefRoleNode) {
  if (!refNode.children || !refNode.children.length) return { ...refNode };
  const copyOfChildren = [...refNode.children];

  const copyOfNode = {
    ...refNode,
    children: copyOfChildren,
  };
  return copyOfNode;
}
