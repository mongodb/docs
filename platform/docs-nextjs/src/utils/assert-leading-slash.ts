const assertLeadingSlash = (url: string) => {
  if (url && url.match(/^\//)) {
    return url;
  }
  return `/${url}`;
};

export { assertLeadingSlash };
