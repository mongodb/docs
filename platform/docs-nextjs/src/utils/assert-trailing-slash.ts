export const assertTrailingSlash = (url: string) => {
  if (url && url.match(/\/$/)) {
    return url;
  }
  return `${url}/`;
};
