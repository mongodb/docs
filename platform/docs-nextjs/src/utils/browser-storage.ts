import { isBrowser } from './is-browser';

const isValidStorage = isBrowser;

export const setLocalValue = (key: string, value: unknown) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      localStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting localStorage value');
  }
};

export const getLocalValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting localStorage value');
    return undefined;
  }
};

export const setSessionValue = (key: string, value: string) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      sessionStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting session value');
  }
};

export const getSessionValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting session value');
    return undefined;
  }
};
