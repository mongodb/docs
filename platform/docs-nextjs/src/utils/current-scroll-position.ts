import { isBrowser } from './is-browser';

// Function that returns current scroll position for analytics
export const currentScrollPosition = (): number => {
  if (isBrowser) {
    const percentage = Math.round((window.scrollY / (document.body.scrollHeight - window.innerHeight)) * 100);
    return Math.round(percentage / 10) * 10;
  }
  return 0;
};
