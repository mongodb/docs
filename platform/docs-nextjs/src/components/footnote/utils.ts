import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
