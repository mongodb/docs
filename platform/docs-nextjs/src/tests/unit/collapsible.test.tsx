import { render, act } from '@testing-library/react';

import { mockLocation } from '../utils/mock-location';
import Collapsible from '@/components/collapsible';
import mockData from '@/tests/data/collapsible.test.json';
import expandedMockData from '@/tests/data/collapsible-expanded.test.json';
import type { CollapsibleNode } from '@/types/ast';

const typedMockData = mockData as CollapsibleNode;
const typedExpandedMockData = expandedMockData as CollapsibleNode;

describe('collapsible component', () => {
  it('renders all the content in the options and children', () => {
    const renderResult = render(
      <Collapsible nodeChildren={typedMockData.children} options={typedMockData.options}></Collapsible>,
    );
    expect(renderResult.asFragment()).toMatchSnapshot();
    expect(renderResult.getByText('This is a heading')).toBeTruthy();
    expect(renderResult.getByText('This is a subheading')).toBeTruthy();
    expect(renderResult.getByText('This is collapsible content')).toBeTruthy();
  });

  it('renders the content and correct icon when interacted', async () => {
    const renderResult = render(
      <Collapsible nodeChildren={typedMockData.children} options={typedMockData.options}></Collapsible>,
    );
    let button = renderResult.getByRole('button');
    const collapsedContent = renderResult.getByText('This is collapsible content');

    let icon = button.querySelector('[role=img]');
    expect(icon?.getAttribute('aria-label')).toContain('Chevron');
    expect(icon?.getAttribute('aria-label')).toContain('Right');

    await act(async () => {
      button.click();
    });

    button = renderResult.getByRole('button');
    icon = button.querySelector('[role=img]');
    expect(icon?.getAttribute('aria-label')).toContain('Chevron');
    expect(icon?.getAttribute('aria-label')).toContain('Down');
    expect(collapsedContent).toBeVisible();
  });

  it('opens the collapsible content if hash is found in the URL', async () => {
    mockLocation({ hash: '#this-is-a-heading' });
    const renderResult = render(
        <Collapsible nodeChildren={typedMockData.children} options={typedMockData.options}></Collapsible>,
      ),
      button = renderResult.getByRole('button'),
      icon = button.querySelector('[role=img]');
    expect(icon?.getAttribute('aria-label')).toContain('Chevron');
    expect(icon?.getAttribute('aria-label')).toContain('Down');
  });

  it('is default expanded if expanded option is truthy', async () => {
    const renderResult = render(
        <Collapsible nodeChildren={typedExpandedMockData.children} options={typedExpandedMockData.options} />,
      ),
      button = renderResult.getByRole('button'),
      icon = button.querySelector('[role=img]');
    expect(icon?.getAttribute('aria-label')).toContain('Chevron');
    expect(icon?.getAttribute('aria-label')).toContain('Down');
  });
});
