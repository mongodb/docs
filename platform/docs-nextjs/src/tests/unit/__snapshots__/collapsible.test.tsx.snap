// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`collapsible component renders all the content in the options and children 1`] = `
<DocumentFragment>
  .emotion-0 {
  border-bottom: 1px solid #E8EDEB;
}

.emotion-0 {
  border-top: 1px solid #e8edeb;
  margin-top: 32px;
}

.emotion-0~.emotion-0 {
  border-top: none;
  margin-top: 0;
}

.emotion-0:nth-child(1 of .collapsible) {
  border-top: 1px solid #E8EDEB;
  margin-top: 32px;
}

.emotion-0:nth-last-child(1 of .collapsible) {
  margin-bottom: 32px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-column-gap: 24px;
  column-gap: 24px;
  margin: 32px 0;
}

.emotion-2 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 24px;
  line-height: 32px;
  font-weight: 500;
  color: #001E2B;
  margin-top: 24px;
  margin-bottom: 8px;
  color: var(--font-color-primary);
  margin-top: 0;
}

.emotion-3 {
  position: absolute;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  padding: 0 10px;
  visibility: hidden;
}

.emotion-4 {
  color: #889397;
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-5 {
  display: none;
}

.emotion-5 {
  display: none;
}

.emotion-6 {
  display: inline;
  left: 0;
  top: 0;
  margin-top: -85px;
  position: absolute;
  padding-bottom: 2px;
}

@media only screen and (max-width: 767px) {
  .emotion-6 {
    margin-top: -85px;
  }
}

.emotion-7 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
}

.emotion-7 strong,
.emotion-7 b {
  font-weight: 600;
}

.emotion-8 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-9 {
  border: none;
  -webkit-appearance: unset;
  padding: unset;
  display: inline-block;
  border-radius: 100px;
  position: relative;
  cursor: pointer;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  transition-property: color,box-shadow;
  background-color: rgba(255, 255, 255, 0);
  height: 28px;
  width: 28px;
  color: #5C6C75;
}

.emotion-9::before {
  content: '';
  -webkit-transition: 150ms all ease-in-out;
  transition: 150ms all ease-in-out;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  border-radius: 100%;
  -webkit-transform: scale(0.8);
  -moz-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
}

.emotion-9:active:before,
.emotion-9:hover:before,
.emotion-9:focus:before,
.emotion-9[data-hover='true']:before,
.emotion-9[data-focus='true']:before {
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.emotion-9:focus {
  outline: none;
}

.emotion-9:active,
.emotion-9:hover,
.emotion-9[data-hover='true'] {
  color: #001E2B;
}

.emotion-9:active::before,
.emotion-9:hover::before,
.emotion-9[data-hover='true']::before {
  background-color: rgba(61,79,88,0.1);
}

.emotion-9:focus-visible,
.emotion-9[data-focus='true'] {
  color: #001E2B;
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC;
}

.emotion-9:focus-visible::before,
.emotion-9[data-focus='true']::before {
  background-color: rgba(61,79,88,0.1);
}

.emotion-10 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-10 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-11 {
  overflow: hidden;
  height: 0;
  color: --font-color-primary;
}

[aria-expanded='true'] .emotion-11 {
  height: auto;
}

.emotion-12 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-12 strong,
.emotion-12 b {
  font-weight: 600;
}

.emotion-13 {
  display: table;
  margin: 24px 0;
  min-width: 150px;
  table-layout: fixed;
  width: 100%;
}

.emotion-13 button>div>div {
  font-size: 13px;
}

.emotion-13 pre {
  background-color: #F9FBFA;
  color: #001E2B;
}

.dark-theme .emotion-13 pre {
  background-color: #001E2B;
  color: #F9FBFA;
}

.emotion-13 [data-testid='leafygreen-code-panel'] {
  background-color: #FFFFFF;
  border-color: #E8EDEB;
}

.dark-theme .emotion-13 [data-testid='leafygreen-code-panel'] {
  background-color: #3D4F58;
  border-color: #3D4F58;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-function,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-class,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-tag,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-attr,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-pseudo,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-id,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-class {
  color: #CC3887;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-function,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-class,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-tag,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-attr,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-pseudo,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-id,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-selector-class {
  color: #FF7DC3;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-regexp,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-number,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-literal,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-function.lg-highlight-title {
  color: #016ee9;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-regexp,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-number,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-literal,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-function.lg-highlight-title {
  color: #2dc4ff;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-quote,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-section,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-name {
  color: #007ab8;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-quote,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-section,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-name {
  color: #a5e3ff;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-string,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-addition {
  color: #12824D;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-string,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-addition {
  color: #35DE7B;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta-string {
  color: #956d00;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta-string {
  color: #EDB210;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-variable,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-deletion,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-symbol,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-bullet,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-link,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-attr,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-attribute,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-language,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-template-variable,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-built_in,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-type,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-params {
  color: #D83713;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-variable,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-deletion,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-symbol,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-bullet,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-link,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-attr,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-attribute,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-language,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-template-variable,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-built_in,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-type,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-params {
  color: #FF6F44;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-title,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-class.lg-highlight-title {
  color: #001E2B;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-title,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-class.lg-highlight-title {
  color: #F9FBFA;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-doctag,
.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-formula {
  color: #001E2B;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-doctag,
.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-formula {
  color: #F9FBFA;
}

.emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-comment {
  color: #3D4F58;
}

.dark-theme .emotion-13 [class*='lg-highlight-hljs-'] .lg-highlight-comment {
  color: #C1C7C6;
}

.emotion-13>div>div {
  display: grid;
  grid-template-columns: code panel;
}

.emotion-13>div {
  border-top-left-radius: 12px;
  border-top-right-radius: 12px;
  border-color: #E8EDEB;
}

.dark-theme .emotion-13>div {
  border-color: #3D4F58;
}

.emotion-14 {
  border: 1px solid #E8EDEB;
  border-radius: 12px;
  overflow: hidden;
}

.emotion-15 {
  position: relative;
  display: grid;
  grid-template-areas: 'code panel';
  grid-template-columns: auto 38px;
  border-radius: inherit;
  z-index: 0;
}

.emotion-15:before,
.emotion-15:after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  height: 100%;
  width: 40px;
  border-radius: 40%;
  box-shadow: unset;
  -webkit-transition: box-shadow 100ms ease-in-out;
  transition: box-shadow 100ms ease-in-out;
}

.emotion-15:before {
  grid-column: 1;
  left: -40px;
}

.emotion-15:after {
  grid-column: 2;
}

.emotion-16 {
  grid-area: code;
  overflow-x: auto;
  border-radius: inherit;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border: 0;
  padding-top: 8px;
  padding-bottom: 8px;
  margin: 0;
  position: relative;
  -webkit-transition: box-shadow 100ms ease-in-out;
  transition: box-shadow 100ms ease-in-out;
  -webkit-transition: height 300ms ease-in-out;
  transition: height 300ms ease-in-out;
  white-space: pre;
  background-color: #F9FBFA;
  color: #001E2B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 6px;
  padding-bottom: 6px;
}

@media only screen and (max-device-width: 812px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-16 {
    white-space: pre-wrap;
  }
}

@media only screen and (min-device-width: 813px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-16 {
    white-space: pre;
  }
}

.emotion-16:focus-visible {
  outline: none;
}

.emotion-17 {
  color: inherit;
  font-family: 'Source Code Pro',Menlo,monospace;
  font-size: 13px;
  line-height: 20px;
}

.emotion-18 {
  border-spacing: 0;
}

.emotion-19 {
  border-spacing: 0;
  vertical-align: top;
  padding: 0 16px;
}

.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  gap: 4px;
  background-color: #FFFFFF;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 6px;
  border-left: solid 1px;
  border-color: #E8EDEB;
  z-index: 2;
  grid-area: panel;
}

.emotion-20 svg {
  width: 16px;
  height: 16px;
}

<div
    aria-expanded="false"
    class="collapsible emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        <h2
          class="contains-headerlink emotion-2"
          weight="medium"
        >
          This is a heading
          <a
            class="headerlink emotion-3"
            href="#this-is-a-heading"
            title="Permalink to this heading"
          >
            <svg
              aria-label="Link Icon"
              class="emotion-4"
              fill="none"
              height="12"
              role="img"
              viewBox="0 0 16 16"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.039 9.953a4 4 0 0 1-.327-.38l1.45-1.445q.11.226.294.414a1.553 1.553 0 0 0 2.203.01l2.884-2.872a1.57 1.57 0 0 0 .01-2.212 1.553 1.553 0 0 0-2.203-.01l-.379.377A.995.995 0 0 1 8.56 3.83a1.005 1.005 0 0 1 .006-1.418l.38-.377a3.54 3.54 0 0 1 5.024.023 3.58 3.58 0 0 1-.022 5.047l-2.884 2.871a3.54 3.54 0 0 1-5.025-.022"
                fill="currentColor"
              />
              <path
                d="M9.961 6.047q.18.181.327.38l-1.45 1.445q-.11-.226-.294-.414a1.553 1.553 0 0 0-2.203-.01L3.457 10.32a1.57 1.57 0 0 0-.01 2.212 1.553 1.553 0 0 0 2.203.01l.379-.377a.995.995 0 0 1 1.411.006 1.005 1.005 0 0 1-.006 1.418l-.38.377a3.54 3.54 0 0 1-5.024-.023 3.58 3.58 0 0 1 .022-5.047l2.884-2.871a3.54 3.54 0 0 1 5.025.022"
                fill="currentColor"
              />
            </svg>
            <span
              class="emotion-5"
            />
            <div
              class="emotion-6"
              id="this-is-a-heading"
            />
          </a>
        </h2>
        <p
          class="emotion-7"
        >
          This is a subheading
        </p>
      </div>
      <button
        aria-disabled="false"
        aria-expanded="false"
        aria-labelledby="Expand the collapsed content"
        class="emotion-8 emotion-9"
        tabindex="0"
      >
        <div
          class="emotion-10"
        >
          <svg
            aria-label="Chevron Right Icon"
            class=""
            fill="none"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.364 14.364a1 1 0 0 0 1.414 0l4.95-4.95.707-.707a1 1 0 0 0 0-1.414l-.707-.707-4.95-4.95a1 1 0 0 0-1.414 0l-.707.707a1 1 0 0 0 0 1.414L8.899 8l-4.242 4.243a1 1 0 0 0 0 1.414z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </button>
    </div>
    <div
      class="emotion-11"
    >
      <p
        class="emotion-12"
      >
        This is collapsible content
      </p>
      <div
        class="intro-code-block emotion-13"
      >
        <div
          class="emotion-14"
        >
          <div
            class="emotion-15"
          >
            <pre
              class="emotion-16"
              tabindex="-1"
            >
              <code
                class="lg-highlight-hljs-light javascript emotion-17"
              >
                <table
                  class="emotion-18"
                >
                  <tbody>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-19"
                      >
                        <span
                          class="lg-highlight-title lg-highlight-class"
                        >
                          This
                        </span>
                         is code within collapsible content
                      </td>
                    </tr>
                  </tbody>
                </table>
              </code>
            </pre>
            <div
              class="emotion-20"
              data-testid="leafygreen-code-panel"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
